src/parser/position.rs

```diff
@@ -0,0 +1,77 @@
+use crate::parser::Span;
+#[cfg(not(target_arch = "wasm32"))]
+use proptest::prelude::*;
+
+/// Source code position of an expression in a file
+#[derive(Clone, Copy, Debug)]
+pub enum Pos {
+    No,
+    Pos {
+        from_offset: usize,
+        from_line: usize,
+        from_column: usize,
+        upto_offset: usize,
+        upto_line: usize,
+        upto_column: usize,
+    },
+}
+
+impl PartialEq for Pos {
+    fn eq(&self, _other: &Self) -> bool {
+        true
+    }
+}
+
+impl Eq for Pos {}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl Arbitrary for Pos {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<()>().prop_map(|_| Pos::No).boxed()
+    }
+}
+
+impl Pos {
+    /// Use the range information in a Position to pretty-print that range within
+    /// a string
+    pub fn range(
+        input: String,
+        from_line: usize,
+        from_column: usize,
+        upto_line: usize,
+        upto_column: usize,
+    ) -> String {
+        let mut res = String::new();
+        let gutter = format!("{}", upto_line).len();
+        let pad = format!("{: >gutter$}", from_line, gutter = gutter).len() + 3 + from_column;
+        res.push_str(&format!("{}▼\n", " ".to_owned().repeat(pad)));
+        for (line_number, line) in input.lines().enumerate() {
+            if ((line_number + 1) >= from_line) && ((line_number + 1) <= upto_line) {
+                res.push_str(&format!(
+                    "{: >gutter$} | {}\n",
+                    line_number + 1,
+                    line,
+                    gutter = gutter
+                ));
+            }
+        }
+        let pad = format!("{: >gutter$}", upto_line, gutter = gutter).len() + 3 + upto_column;
+        res.push_str(&format!("{}▲", " ".to_owned().repeat(pad)));
+        res
+    }
+
+    /// Construct a position from the difference of two Spans
+    pub fn from_upto(from: Span<'_>, upto: Span<'_>) -> Self {
```

Comment:
Is this used?

Reply:
yes, it's how we annotate the AST node with the start and end source positions that generated it:

```
pub fn parse_string<F: LurkField>(
) -> impl Fn(Span<'_>) -> IResult<Span<'_>, Syntax<F>, ParseError<Span<'_>, F>> {
    move |from: Span<'_>| {
        let (upto, s) = string::parse_string('"')(from)?;
        let pos = Pos::from_upto(from, upto);
        Ok((upto, Syntax::String(pos, s)))
    }
}
```